pkg/streamingpromql/operators/aggregation.go

```diff
@@ -131,58 +152,52 @@ func (a *Aggregation) NextSeries(ctx context.Context) (types.InstantVectorSeries
 		return types.InstantVectorSeriesData{}, types.EOS
 	}
 
-	start := timestamp.FromTime(a.Start)
-	end := timestamp.FromTime(a.End)
-	interval := a.Interval.Milliseconds()
-	steps := stepCount(start, end, interval)
-
 	// Determine next group to return
 	thisGroup := a.remainingGroups[0]
 	a.remainingGroups = a.remainingGroups[1:]
 
 	// Iterate through inner series until the desired group is complete
-	for thisGroup.remainingSeriesCount > 0 {
-		s, err := a.Inner.NextSeries(ctx)
+	if err := a.accumulateUntilGroupComplete(ctx, thisGroup); err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	// Construct the group and return it
+	seriesData, err := a.constructSeriesData(thisGroup, a.start, a.interval)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	groupPool.Put(thisGroup)
+	return seriesData, nil
+}
+
+func (a *Aggregation) accumulateUntilGroupComplete(ctx context.Context, g *group) (err error) {
```

Comment:
[nit] The return value name isn't needed anymore

```suggestion
func (a *Aggregation) accumulateUntilGroupComplete(ctx context.Context, g *group) error {
```

Same applies to `accumulateSeriesIntoGroup` below as well.

Reply:
It is in accumulateSeriesIntoGroup if we want to avoid `var err error`, but I'll change it for consistency